File: backend/app/services/guards.py
"""Write safety guards for RouterOS mutations.

Prevents self-lockout, accidental throttling of management endpoints,
pruning of foreign WinBox queues/rules, and invalid rate configurations.
"""
import re
from typing import Optional, Set

RATE_PATTERN = re.compile(r"^(\d+)([kKMGT]?)/(\d+)([kKMGT]?)$")
MULTIPLIERS = {
    "": 1,
    "k": 1_000,
    "K": 1_000,
    "M": 1_000_000,
    "m": 1_000_000,
    "G": 1_000_000_000,
    "g": 1_000_000_000,
    "T": 1_000_000_000_000,
    "t": 1_000_000_000_000,
}

IMMUNE_WILDCARDS = {"0.0.0.0", "0.0.0.0/0", "::/0", "255.255.255.255"}
IMMUNE_LOOPBACKS = {"127.0.0.1", "::1", "localhost"}


class WriteGuardViolation(ValueError):
    """Raised when an operation is refused by a RouterOS write safety guard."""

    def __init__(self, guard_name: str, reason: str, target: str):
        super().__init__(f"[WriteGuard] [{guard_name}] Refused write for {target}: {reason}")
        self.guard_name = guard_name
        self.reason = reason
        self.target = target


def parse_bps(val: str) -> int:
    """Parse bandwidth rate like '5M', '100k', or '0' into bits per second."""
    raw = str(val).strip()
    if raw.isdigit():
        return int(raw)
    match = re.match(r"^(\d+)([kKMGT]?)$", raw)
    if not match:
        raise ValueError(f"Invalid rate string: {val}")
    num, unit = match.groups()
    return int(num) * MULTIPLIERS.get(unit, 1)


def parse_pair(pair_str: str) -> tuple[int, int]:
    """Parse a pair rate like '5M/10M' or '0/0' into (upload_bps, download_bps)."""
    clean = str(pair_str).strip()
    if clean in ("0", "0/0", "unlimited", "none"):
        return 0, 0
    match = RATE_PATTERN.match(clean)
    if not match:
        raise ValueError(f"Invalid rate pair format: {pair_str}")
    up_num, up_unit, down_num, down_unit = match.groups()
    return (
        int(up_num) * MULTIPLIERS.get(up_unit, 1),
        int(down_num) * MULTIPLIERS.get(down_unit, 1),
    )


def guard_immune_targets(target: str, immune_ips: Set[str], action: str = "block") -> None:
    """Refuse blocking or throttling of immune infrastructure and host targets."""
    clean_target = str(target).strip()
    ip_only = clean_target.split("/")[0]

    if ip_only in IMMUNE_LOOPBACKS or clean_target in IMMUNE_LOOPBACKS:
        raise WriteGuardViolation(
            "ImmuneTargetGuard",
            "Target is a loopback address and cannot be modified",
            clean_target,
        )

    if clean_target in IMMUNE_WILDCARDS:
        raise WriteGuardViolation(
            "ImmuneTargetGuard",
            "Target is a wildcard/broadcast and cannot be throttled or blocked",
            clean_target,
        )

    if ip_only in immune_ips or clean_target in immune_ips:
        raise WriteGuardViolation(
            "ImmuneTargetGuard",
            f"Target {clean_target} is a protected management/host IP",
            clean_target,
        )


def guard_foreign_resources(comment: Optional[str], action: str, resource_type: str) -> None:
    """Refuse destructive operations on resources not managed by MikroMan."""
    c = (comment or "").strip()
    if not c.startswith("mikroman:"):
        raise WriteGuardViolation(
            "ForeignResourceGuard",
            f"Cannot {action} foreign {resource_type} without 'mikroman:' comment prefix",
            c or "<empty>",
        )


def guard_queue_invariants(
    target: str,
    max_limit: str,
    limit_at: Optional[str] = None,
    parent: Optional[str] = None,
    name: Optional[str] = None,
) -> None:
    """Verify queue parameters comply with RouterOS relational requirements."""
    try:
        max_up, max_down = parse_pair(max_limit)
    except ValueError as e:
        raise WriteGuardViolation("QueueInvariantGuard", str(e), target)

    if limit_at:
        # Only the parse is wrapped. WriteGuardViolation subclasses ValueError,
        # so a violation raised inside a `try` whose `except` catches ValueError
        # would be caught and re-wrapped, nesting the message inside itself.
        try:
            at_up, at_down = parse_pair(limit_at)
        except ValueError as e:
            raise WriteGuardViolation("QueueInvariantGuard", str(e), target)

        if max_up > 0 and at_up > max_up:
            raise WriteGuardViolation(
                "QueueInvariantGuard",
                f"Upload limit-at ({at_up} bps) cannot exceed max-limit ({max_up} bps)",
                target,
            )
        if max_down > 0 and at_down > max_down:
            raise WriteGuardViolation(
                "QueueInvariantGuard",
                f"Download limit-at ({at_down} bps) cannot exceed max-limit ({max_down} bps)",
                target,
            )

    if parent and name and parent.strip() == name.strip():
        raise WriteGuardViolation(
            "QueueInvariantGuard",
            f"Queue cannot be its own parent: circular parentage on {name}",
            target,
        )

File: backend-go/internal/routeros/guards.go
package routeros

import (
	"fmt"
	"net"
	"regexp"
	"strconv"
	"strings"
)

var (
	ratePattern       = regexp.MustCompile(`^(\d+)([kKMGT]?)/(\d+)([kKMGT]?)$`)
	rateSinglePattern = regexp.MustCompile(`^(\d+)([kKMGT]?)$`)

	immuneWildcards = map[string]bool{
		"0.0.0.0":         true,
		"0.0.0.0/0":       true,
		"::/0":            true,
		"255.255.255.255": true,
	}

	immuneLoopbacks = map[string]bool{
		"127.0.0.1": true,
		"::1":       true,
		"localhost": true,
	}
)

type WriteGuardViolation struct {
	Guard  string
	Target string
	Reason string
}

func (e *WriteGuardViolation) Error() string {
	return fmt.Sprintf("[WriteGuard] [%s] Refused write for %s: %s", e.Guard, e.Target, e.Reason)
}

// ParseBps parses bandwidth rate like "5M", "100k", or "0" into bits per second.
func ParseBps(val string) (int64, error) {
	raw := strings.TrimSpace(val)
	if raw == "" || raw == "0" {
		return 0, nil
	}
	if n, err := strconv.ParseInt(raw, 10, 64); err == nil {
		return n, nil
	}
	m := rateSinglePattern.FindStringSubmatch(raw)
	if m == nil {
		return 0, fmt.Errorf("invalid rate string: %s", val)
	}
	num, _ := strconv.ParseInt(m[1], 10, 64)
	mult := unitMultiplier(m[2])
	return num * mult, nil
}

// ParseRatePair parses "5M/10M" or "0/0" into (upload_bps, download_bps).
func ParseRatePair(pair string) (int64, int64, error) {
	clean := strings.TrimSpace(pair)
	if clean == "" || clean == "0" || clean == "0/0" || clean == "unlimited" || clean == "none" {
		return 0, 0, nil
	}
	m := ratePattern.FindStringSubmatch(clean)
	if m == nil {
		return 0, 0, fmt.Errorf("invalid rate pair format: %s", pair)
	}
	upNum, _ := strconv.ParseInt(m[1], 10, 64)
	upMult := unitMultiplier(m[2])
	downNum, _ := strconv.ParseInt(m[3], 10, 64)
	downMult := unitMultiplier(m[4])
	return upNum * upMult, downNum * downMult, nil
}

func unitMultiplier(unit string) int64 {
	switch unit {
	case "k", "K":
		return 1_000
	case "m", "M":
		return 1_000_000
	case "g", "G":
		return 1_000_000_000
	case "t", "T":
		return 1_000_000_000_000
	default:
		return 1
	}
}

// GuardImmuneTarget ensures management endpoints, wildcards, loopbacks, and supernets containing them are never blocked or throttled.
func GuardImmuneTarget(target string, immunes map[string]bool, action string) error {
	clean := strings.TrimSpace(target)
	if clean == "" {
		return nil
	}

	// 1. Loopbacks
	if immuneLoopbacks[clean] {
		return &WriteGuardViolation{Guard: "immune_targets", Target: target, Reason: "loopback targets are immune"}
	}
	if strings.HasPrefix(clean, "127.") {
		return &WriteGuardViolation{Guard: "immune_targets", Target: target, Reason: "loopback targets are immune"}
	}

	// 2. Wildcards
	if immuneWildcards[clean] {
		return &WriteGuardViolation{Guard: "immune_targets", Target: target, Reason: "wildcard targets are immune"}
	}

	// 3. Direct immune matches
	ipOnly := cleanIP(clean)
	if immunes[clean] || immunes[ipOnly] {
		return &WriteGuardViolation{Guard: "immune_targets", Target: target, Reason: "protected router management target"}
	}

	// 4. Supernet containment: if target is a CIDR network that contains an immune IP
	if strings.Contains(clean, "/") {
		_, ipNet, err := net.ParseCIDR(clean)
		if err == nil && ipNet != nil {
			for imm := range immunes {
				parsedImm := net.ParseIP(imm)
				if parsedImm != nil && ipNet.Contains(parsedImm) {
					return &WriteGuardViolation{
						Guard:  "immune_targets",
						Target: target,
						Reason: fmt.Sprintf("supernet contains immune address %s", imm),
					}
				}
			}
		}
	}

	return nil
}

// GuardForeignResource ensures we never mutate or delete rules/queues not created by MikroMan.
func GuardForeignResource(comment string, action, resourceType string) error {
	if !strings.HasPrefix(comment, "mikroman:") {
		return &WriteGuardViolation{
			Guard:  "foreign_resources",
			Target: comment,
			Reason: fmt.Sprintf("foreign %s without 'mikroman:' prefix is protected", resourceType),
		}
	}
	return nil
}

// GuardQueueInvariants validates queue rate formatting, hierarchy, and limits.
func GuardQueueInvariants(target, maxLimit, limitAt, parent, name string) error {
	if parent != "" && parent != "none" && parent == name {
		return &WriteGuardViolation{
			Guard:  "queue_invariants",
			Target: target,
			Reason: fmt.Sprintf("circular parent reference: %s cannot be parent of itself", name),
		}
	}

	upMax, downMax, err := ParseRatePair(maxLimit)
	if err != nil {
		return &WriteGuardViolation{
			Guard:  "queue_invariants",
			Target: target,
			Reason: fmt.Sprintf("invalid max_limit format: %s", maxLimit),
		}
	}

	if limitAt != "" && limitAt != "0" && limitAt != "0/0" && limitAt != "none" {
		upAt, downAt, err := ParseRatePair(limitAt)
		if err != nil {
			return &WriteGuardViolation{
				Guard:  "queue_invariants",
				Target: target,
				Reason: fmt.Sprintf("invalid limit_at format: %s", limitAt),
			}
		}

		if upMax > 0 && upAt > upMax {
			return &WriteGuardViolation{
				Guard:  "queue_invariants",
				Target: target,
				Reason: fmt.Sprintf("upload limit_at (%s) cannot exceed max_limit (%s)", limitAt, maxLimit),
			}
		}
		if downMax > 0 && downAt > downMax {
			return &WriteGuardViolation{
				Guard:  "queue_invariants",
				Target: target,
				Reason: fmt.Sprintf("download limit_at (%s) cannot exceed max_limit (%s)", limitAt, maxLimit),
			}
		}
	}

	return nil
}
